anndata/_config.py

```diff
@@ -0,0 +1,332 @@
+        if isinstance(option, Iterable) and not isinstance(option, str):
+            return "\n".join(
+                [self.describe(k, print_description=print_description) for k in option]
+            )
+        doc = self._registered_options[option].doc
+        if option in self._deprecated_options:
+            doc += "\n"
+            opt = self._deprecated_options[option]
+            if opt.message is not None:
+                doc += opt.message + "\n"
+            doc += f"{option} will be removed in {opt.removal_version}"
+        if print_description:
+            print(doc)
+        return doc
+
+    def deprecate(
+        self, option: str, removal_version: str, message: str | None = None
+    ) -> None:
+        """Deprecate options with a message at a version.
+
+        Parameters
+        ----------
+        option
+            Which option should be deprecated.
+        removal_version
+            The version targeted for removal.
+        message
+            A custom message.
+        """
+        self._deprecated_options[option] = DeprecatedOption(
+            option, message, removal_version
+        )
+
+    @update_override_function
+    def register(
+        self,
+        option: str,
+        default_value: object,
+        doc: str,
+        validator: Callable[[T], None],
+    ) -> None:
+        """Register an option so it can be set/described etc. by end-users
+
+        Parameters
+        ----------
+        option
+            Option to be set.
+        default_value
+            Default value with which to set the option.
+        doc
+            Docstring for the option
+        validator
+            A function which asserts that the option's value is valid.
+        """
+        validator(default_value)
+        self._registered_options[option] = RegisteredOption(
+            option, default_value, doc, validator
+        )
+        self._config[option] = default_value
+
+    def __setattr__(self, option: str, val: object) -> None:
+        """
+        Set an option to a value.  To see the allowed option to be set and their description,
+        use describe_option.
+
+        Parameters
+        ----------
+        option
+            Option to be set.
+        val
+            Value with which to set the option.
+
+        Raises
+        ------
+        KeyError
+            If the option has not been registered, this function will raise an error.
+        """
+        if hasattr(super(), option):
+            super().__setattr__(option, val)
+        elif option not in self._registered_options:
+            raise KeyError(
+                f"{option} is not an available option for anndata.\
+                Please open an issue if you believe this is a mistake."
+            )
+        else:
+            registered_option = self._registered_options[option]
+            registered_option.validator(val)
+            self._config[option] = val
+
+    def __getattr__(self, option: str) -> object:
```

Comment:
I guess that’s the only point where missing static typing is painful.

Using `settings.foo` anywhere in the codebase will only be `object`

Reply:
So yes, I think you're right.  I had not checked in VSCode for the attribute completion (only IPython kernels).  VSCode's type checker is static so this is not going to happen.  This seems like a deal breaker if I am not mistaken?

Reply:
> This seems like a deal breaker if I am not mistaken?

depends on if we expect the settings mechanism to be used from library code or almost exclusively from interactive code.